experimental/wgit/weigit_api.py

```diff
@@ -0,0 +1,141 @@
+            os.mkdir(".wgit")
+        except FileExistsError:
+            sys.stderr.write("WeiGit has been already Initialized \n")
+            sys.exit(0)
+
+        # if no .wgit dir then initialize the following
+        SHA1_store()
+
+        # create sha1_ref_count
+        with open(".wgit/sha1_ref_count.json", "w") as f:
+            print("The sha1_ref_count.json file has been created!")
+
+        # Make the .wgit a git repo, add a gitignore and add SHA1_ref
+        pygit2.init_repository(".wgit/.git", False)
+        with open("./.wgit/.gitignore", "w") as f:
+            f.write("sha1_ref_count.json")
+
+    @staticmethod
+    def add(file):
+        if Repo(os.getcwd()).exists():
+            print("wgit added")
+
+    @staticmethod
+    def status():
+        if Repo(os.getcwd()).exists():
+            print("wgit status")
+
+    @staticmethod
+    def log(file):
+        if Repo(os.getcwd()).exists():
+            if file:
+                print(f"wgit log of the file {file}")
+            else:
+                print("wgit log")
+
+    @staticmethod
+    def commit(message):
+        if Repo(os.getcwd()).exists():
+            if message:
+                print(f"commited with message: {message}")
+            else:
+                print("wgit commit")
+
+    @staticmethod
+    def checkout():
+        if Repo(os.getcwd()).exists():
+            print("wgit checkout")
+
+    @staticmethod
+    def compression():
+        print("Not Implemented!")
+
+    @staticmethod
+    def checkout_by_steps():
+        print("Not Implemented!")
+
+
+class SHA1_store:
+    """
+    Planned Features:
+        1. def init
+        2. def add <file or data> -> SHA1
+        3. def remove (SHA1)
+        4. def add_ref(children_SHA1, parent_SHA1)
+        5. def read(SHA1): ->
+        6. def lookup(SHA1): -> file path to the data. NotFound Exception if not found.
+    """
+
+    def __init__(self) -> None:
+        print("SHA1_store has been initialized!!")
+
+
+class Repo:
+    """
+    Designates the weigit repo, which is identified by a path to the repo.
+    """
+
+    def __init__(self, check_dir) -> None:
+        self.repo_path = None
+        self.check_dir = check_dir
+
+    def exists(self):
+        def weigit_repo_exists(check_dir):
+            """
+            checks if the input path to dir (check_dir) is a valid weigit repo
+            with .git and sha1_ref_count in the repo.
+            """
+            is_wgit_in_curr = pathlib.Path(os.path.join(check_dir, ".wgit")).exists()
+            is_refcount_in_wgit = pathlib.Path(os.path.join(check_dir, ".wgit/sha1_ref_count.json")).exists()
+            is_git_in_wgit = pathlib.Path(os.path.join(check_dir, ".wgit/.git")).exists()
+            return is_wgit_in_curr and is_refcount_in_wgit and is_git_in_wgit
+
+        if weigit_repo_exists(self.check_dir):
+            self.repo_path = os.path.join(self.check_dir, ".wgit")
+        else:
+            max_dir_depth = 100
```

Comment:
add a comment on the need of this? max_dir_depth is used to cache cycles in the symbolic links? Have you tried os.path.realpath() to get the real path for a given path?

Reply:
The `max_dir_depth` is simply so that we do not get into an infinite while loop. The while loop basically starts from the input directory and keeps looping one dir up till it reaches the root dir. Could you suggest a better way to handle the loop or will a comment suffice for this magic number? We can probably error out as well.

In terms of os.path.realpath, are you suggesting parsing the constructor input path with os realpath at the start in the constructor to strip of any sym links?

Reply:
I am wondering when will we get into an infinite loop? Have you run into it? there shouldn't be any cycle in the path, which means it should always go to root. I was hoping that either infinite loop is either not possible or can be dealt with realpath.

Reply:
To be honest, we will probably not get into an infinite loop as you imply. We can probably discard it.  Also, since we are going up the directory linearly, I think we can simply get the realpath of the input dir path at the start. Making these changes.